test/e2e/network/netpol/network_policy.go

```diff
@@ -0,0 +1,884 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package netpol
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/kubernetes/test/e2e/network"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/onsi/ginkgo"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+const (
+	addSCTPContainers  = false
+	isVerbose          = true
+	useFixedNamespaces = false
+
+	// TODO resolve semantics of loopback network policies: should they always be ALLOWED;
+	//   how do Services affect this?  Calico, Cillium, Antrea seem to do different things.
+	ignoreLoopback = true
+)
+
+/*
+You might be wondering, why are there multiple namespaces used for each test case?
+
+These tests are based on "truth tables" that compare the expected and actual connectivity of each pair of pods.
+Since network policies live in namespaces, and peers can be selected by namespace,
+howing the connectivity of pods in other namespaces is key information to show whether a network policy is working as intended or not.
+
+We use 3 namespaces each with 3 pods, and probe all combinations ( 9 pods x 9 pods = 81 data points ) -- including cross-namespace calls.
+
+Here's an example of a test run, showing the expected and actual connectivity, along with the differences.  Note how the
+visual representation as a truth table greatly aids in understanding what a network policy is intended to do in theory
+and what is happening in practice:
+
+		Oct 19 10:34:16.907: INFO: expected:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	X	.	.	.	.	.	.	.	.
+		x/b	X	.	.	.	.	.	.	.	.
+		x/c	X	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	X	.	.	.	.	.	.	.	.
+		z/b	X	.	.	.	.	.	.	.	.
+		z/c	X	.	.	.	.	.	.	.	.
+
+		Oct 19 10:34:16.907: INFO: observed:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	X	.	.	.	.	.	.	.	.
+		x/b	X	.	.	.	.	.	.	.	.
+		x/c	X	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	X	.	.	.	.	.	.	.	.
+		z/b	X	.	.	.	.	.	.	.	.
+		z/c	X	.	.	.	.	.	.	.	.
+
+		Oct 19 10:34:16.907: INFO: comparison:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	.	.	.	.	.	.	.	.	.
+		x/b	.	.	.	.	.	.	.	.	.
+		x/c	.	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	.	.	.	.	.	.	.	.	.
+		z/b	.	.	.	.	.	.	.	.	.
+		z/c	.	.	.	.	.	.	.	.	.
+*/
+var _ = network.SIGDescribe("Netpol [LinuxOnly]", func() {
+	f := framework.NewDefaultFramework("netpol")
+
+	shouldCreateRandomPolicies := true
+	ginkgo.BeforeEach(func() {
+		if shouldCreateRandomPolicies {
+			if useFixedNamespaces {
+				initializeResources(f)
+			}
+			for _, policy := range GetRandomIngressPolicies(21) {
```

Comment:
21 should be a constant with an explanation?

Reply:
imo we don't need this anymore now that we are supporting parallel test runs.  I'll remove it

Reply:
(removed), resolving